### PR TITLE
some async cleanup + pvcProtection as filter

### PR DESCRIPTION
**Pull request by dbenque**
Moving some cleanup to async processing.

Moving the PV Protection as Filter

### CODE REVIEW

File: cmd/draino/draino.go [hunk 1]
```diff
@@ -541,9 +543,8 @@ func controllerRuntimeBootstrap(options *Options, cfg *controllerruntime.Config,
 		drain_runner.WithRetryWall(retryWall),
 		drain_runner.WithLogger(mgr.GetLogger()),
 		drain_runner.WithSharedIndexInformer(indexer),
-		drain_runner.WithPVProtector(pvProtector),
 		drain_runner.WithEventRecorder(eventRecorder),
-		drain_runner.WithFilter(filterFactory.BuildScopeFilter()),
+		drain_runner.WithFilter(filterFactory.BuildCandidateFilter()),
```
Comment (dbenque) [inline cmd/draino/draino.go:547]: We can use `BuildCandidateFilter` in all the places. All the filtering done for candidate also make sense to stop (not starting) draining a done.

File: internal/candidate_runner/info.go [hunk 2]
```diff
@@ -9,20 +9,27 @@ import (
...
 	FilteredOutCount   int
 	Slots              string
 	ProcessingDuration time.Duration
 	LastTime           time.Time
-
 	// private filed that should not go through the serialization
 	lastNodeIterator scheduler.ItemProvider[*v1.Node]
 }
 
+type DataInfoCleanup struct {
```
Comment (dbenque) [inline internal/candidate_runner/info.go:27]: Not yet consumed in the CLI. Will do in another PR if needed.

File: internal/drain_runner/fake.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ type FakeOptions struct {
 	Preprocessors []DrainPreProzessor
 	Logger        *logr.Logger
 	RerunEvery    time.Duration
-	PVProtector   protector.PVProtector
```
Comment (dhenkel92) [inline internal/drain_runner/fake.go:27]: I think we can remove it completely from the fake factory

File: internal/candidate_runner/runner.go [hunk 4]
```diff
@@ -91,17 +91,12 @@ func (runner *candidateRunner) Run(info *groups.RunnerInfo) error {
 
 		// TODO add metric to track amount of nodes in the group
 		if len(nodes) == 0 {
-			// If there are no candidates left, we'll stop the loop
+			// If there are no candidates left, we'll stop the loop and the runner by closing the context
```
Comment (dhenkel92) [inline internal/candidate_runner/runner.go:94]: This part of the code isn't aware of any runner. The outside logic might change in the future

File: internal/candidate_runner/runner.go [hunk 5]
```diff
@@ -200,3 +188,40 @@ func (runner *candidateRunner) handleRetryFlagOnNodes(ctx context.Context, nodes
...
+
+// runCleanupWithContext perform cleanup activities on nodes of the group
+// - handleRetryFlagOnNodes
+// -
+func (runner *candidateRunner) runCleanupWithContext(ctx context.Context, info *groups.RunnerInfo) {
+	// start the cleanup shifted compare to main runner to spread CPU consumption
+	time.Sleep(runner.runEvery / 2)
+	// run an endless loop until there are no drain candidates left
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
+		span, ctx := tracer.StartSpanFromContext(ctx, "runCleanupWithContext")
```
Comment (dhenkel92) [inline internal/candidate_runner/runner.go:200]:
```suggestion
		span, ctx := tracer.StartSpanFromContext(ctx, "RunCleanupWithContext")
```

File: internal/candidate_runner/info.go (hunk 2 above)
Comment (dhenkel92) [reply internal/candidate_runner/info.go:27]: The name is a little bit confusing. Before reading the rest of the code I thought that this is somehow used to cleanup the DataInfo struct 😅

File: cmd/draino/draino.go (hunk 1 above)
Comment (dhenkel92) [reply cmd/draino/draino.go:547]: What changed your mind?
https://github.com/DataDog/draino/pull/128#discussion_r1058213426

File: internal/drain_runner/runner.go [hunk 6]
```diff
@@ -71,6 +67,8 @@ func (runner *drainRunner) Run(info *groups.RunnerInfo) error {
 			info.Data.Set(DrainRunnerInfo, drainInfo)
 		}()
 
+		// Can't be done asynchronously, must be done in sequence with `handleGroup/handleCandidate` because these
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:70]: hmm.. you are right. We might remove the taint during an actual drain 😞

File: internal/candidate_runner/runner.go (hunk 5 above)
Comment (dbenque) [reply internal/candidate_runner/runner.go:200]: Done.

File: cmd/draino/draino.go (hunk 1 above)
Comment (dbenque) [reply cmd/draino/draino.go:547]: I tried to think about possible side effect and finally found none.
More than that reapplying the same set of filter completely make sense: there is a delay between the candidate_runnner and the drain_runner that handle the candidate. That delay can be short (few ms) or longer (couple of seconds/minutes if we have saturation or long lasting drain). To be consistent we need to be sure that same predicates are applied just before the drain. Having the same set of filters is the only way to go if we want to guarantee this consistency.
So we have to apply the same filters to enter into the candidate taint and before transitioning to the draining taint. After that point we should not go back, once the drain is engage we have to continue till its complete success or failure. We cannot reuse the filters again once entered into draining phase. For example the stability-period filter would always trigger due to the current drain activity.

File: internal/candidate_runner/info.go (hunk 2 above)
Comment (dbenque) [reply internal/candidate_runner/info.go:27]: I see your point, I propose to rename to DataInfoForCleanupActivity

File: internal/kubernetes/util.go [hunk 7]
```diff
@@ -335,6 +348,12 @@ func PatchDeleteNodeLabelKey(ctx context.Context, kclient kubernetes.Interface,
 	return PatchNode(ctx, kclient, nodeName, annotationDeletePatch)
 }
 
+func PatchDeleteNodeLabelKeyCR(ctx context.Context, client client.Client, node *core.Node, key string) error {
```
Comment (dhenkel92) [inline internal/kubernetes/util.go:351]:
```suggestion
func PatchDeleteNodeAnnotationKeyCR(ctx context.Context, client client.Client, node *core.Node, key string) error {
```

File: internal/candidate_runner/runner.go [hunk 8]
```diff
@@ -189,9 +177,16 @@ func (runner *candidateRunner) handleRetryFlagOnNodes(ctx context.Context, nodes
...
-			if _, err := runner.retryWall.ResetRetryCount(ctx, node); err != nil {
-				errors = append(errors, fmt.Errorf("cannot reset retry wall on node '%s': %v", node.Name, err))
+			if runner.retryWall.GetDrainRetryAttemptsCount(node) != 0 {
+				node, err = runner.retryWall.ResetRetryCount(ctx, node)
+				if err != nil {
+					errors = append(errors, fmt.Errorf("cannot reset retry wall on node '%s': %v", node.Name, err))
+				}
+			}
+			if errAnnotation := kubernetes.PatchDeleteNodeLabelKeyCR(ctx, runner.client, node, drainRetryFailedAnnotationKey); errAnnotation != nil {
+				errors = append(errors, fmt.Errorf("cannot reset retry wall on node '%s': %v", node.Name, errAnnotation))
```
Comment (dhenkel92) [inline internal/candidate_runner/runner.go:189]:
```suggestion
				errors = append(errors, fmt.Errorf("cannot remove retry restart annotation from node '%s': %v", node.Name, errAnnotation))
```

**Review by dhenkel92 — APPROVED** [review]
